hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/utils/Pipelines.java

```diff
@@ -450,4 +473,159 @@ public static class DummySink implements SinkFunction<Object> {
+    }
+
+    @Override
+    public void invoke(Object value, SinkFunction.Context context) throws Exception {
+      String partition = (String) value;
+      activePartitions.add(partition);
+      long watermark = context.currentWatermark();
+      Iterator<String> it = activePartitions.iterator();
+      while (it.hasNext()) {
+        String partitionPath = it.next();
+        // Convert the partition path to timestamp if the table is partitioned by time field, like day, hour
+        Long partitionTimestamp = convertTimestampByPartitionPath(partitionPath);
+        // If the watermark is greater than the partition timestamp plus the delay time, it represents the
+        // minimum timestamp in the streaming data is beyond the partition max timestamp, so add the partition
+        // path to the finished partitions set and remove it from active partitions set.
+        if (partitionTimestamp + partitionSuccessFileDelay.toMillis() < watermark) {
+          synchronized (checkpointLock) {
+            finishedPartitions.add(partitionPath);
+            it.remove();
+          }
+        }
+      }
+    }
+
+    @Override
+    public void snapshotState(FunctionSnapshotContext functionSnapshotContext) throws Exception {
+      synchronized (checkpointLock) {
+        // Save the partition path of active & finished partition path to state.
+        Preconditions.checkNotNull(activePartitions);
+        Preconditions.checkNotNull(finishedPartitions);
+        activePartitionsState.update(new ArrayList<>(activePartitions));
+        finishedPartitionsState.update(new ArrayList<>(finishedPartitions));
+      }
+    }
+
+    @Override
+    public void initializeState(FunctionInitializationContext context) throws Exception {
+      ListStateDescriptor<String> activePartitionStateDesc =
+              new ListStateDescriptor<>(ACTIVE_PARTITION_STATE_NAME, String.class);
+      activePartitionsState = context.getOperatorStateStore().getListState(activePartitionStateDesc);
+      ListStateDescriptor<String> finishedPartitionStateDesc =
+              new ListStateDescriptor<>(FINISHED_PARTITION_STATE_NAME, String.class);
+      finishedPartitionsState = context.getOperatorStateStore().getListState(finishedPartitionStateDesc);
+      if (context.isRestored()) {
+        activePartitions = new TreeSet<>();
+        finishedPartitions = new TreeSet<>();
+        for (String p : activePartitionsState.get()) {
+          activePartitions.add(p);
+        }
+        for (String p : finishedPartitionsState.get()) {
+          finishedPartitions.add(p);
+        }
+      } else {
+        activePartitions = new TreeSet<>();
+        finishedPartitions = new TreeSet<>();
+      }
+    }
+
+    @Override
+    public void notifyCheckpointComplete(long l) throws Exception {
+      Iterator<String> it = finishedPartitions.iterator();
+      //Iterate the finished partitions set, and write success file to the path of partition.
+      synchronized (checkpointLock) {
```

Comment:
Or you can base on AtomicBoolean
Would it be better to implement CAS in the way of ReentrantLock to judge whether the current writer holds a lock?

for example:
```
if (atomicBoolean.compareAndSet(false, true)) {
	checkpointLock= new Object();
}
```
I'm worried that the lock granularity might be too large here

Reply:
OK,  the checkpoint lock was removed.